File: app/src/main/java/com/android_cademy/starwarstinder/di/AppModule.java
package com.android_cademy.starwarstinder.di;

import android.app.Application;
import androidx.room.Room;
import android.content.Context;

import com.android_cademy.starwarstinder.ViewModel.MainViewModelFactory;
import com.android_cademy.starwarstinder.db.AppDatabase;
import com.android_cademy.starwarstinder.model.ProfileRepository;
import com.android_cademy.starwarstinder.network.ProfileNetworkDataSource;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class AppModule {

    private Application application;

    public AppModule(Application application) {
        this.application = application;
    }

    @Provides
    @Singleton
    Context providesAppContext() {
        return application.getApplicationContext();
    }

    @Provides
    @Singleton
    AppDatabase providesDatabase(Context context) {
        return Room.inMemoryDatabaseBuilder(context.getApplicationContext(), AppDatabase.class)
                .build();
    }

    @Provides
    @Singleton
    ProfileNetworkDataSource providesProfileNetworkSource(Context context) {
        return new ProfileNetworkDataSource(context);
    }

    @Provides
    @Singleton
    ProfileRepository providesProfileRepository(AppDatabase appDatabase,
                                                ProfileNetworkDataSource profileNetworkDataSource) {
        return new ProfileRepository(appDatabase, profileNetworkDataSource);
    }

    @Provides
    @Singleton
    MainViewModelFactory providesMainActivityViewModel(
            ProfileRepository profileRepository) {
        return new MainViewModelFactory(profileRepository);
    }
}

File: app/src/main/java/com/android_cademy/starwarstinder/di/AppModule.kt
package com.android_cademy.starwarstinder.di

import android.app.Application
import android.content.Context
import com.android_academy.db.AppDatabase
import com.android_academy.db.DbBuilder
import com.android_cademy.starwarstinder.model.ProfileRepository
import com.android_academy.network.ProfileNetworkDataSource
import com.android_cademy.starwarstinder.viewModel.MainViewModelFactory
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
class AppModule(private val application: Application) {

    @Provides
    @Singleton
    fun providesAppContext(): Context {
        return application.applicationContext
    }

    @Provides
    @Singleton
    fun providesDatabase(context: Context): AppDatabase {
        return DbBuilder.Builder(context).build()
    }

    @Provides
    @Singleton
    fun providesProfileNetworkSource(context: Context): ProfileNetworkDataSource {
        return ProfileNetworkDataSource(context)
    }

    @Provides
    @Singleton
    fun providesProfileRepository(appDatabase: AppDatabase,
                                  profileNetworkDataSource: ProfileNetworkDataSource): ProfileRepository {
        return ProfileRepository(appDatabase, profileNetworkDataSource)
    }

    @Provides
    @Singleton
    fun providesMainActivityViewModel(
            profileRepository: ProfileRepository): MainViewModelFactory {
        return MainViewModelFactory(profileRepository)
    }
}
